### PR TITLE
Bug fixes and improvements for the udev test suite

### PR DESCRIPTION
**Pull request by mwilck**
This patch set fixes a few minor problems and adds new features to the udev test program, udev-test.pl, and the python code that is used to set up the fake sysfs for these tests.
The new features are:
 - allow addition and removal of several devices in a single test
 - allow testing for the (non-)existence of several symlinks per device in a single test
 - allow auto-generation of test device lists from the fake sysfs
 - generate (almost) arbitrarily many SCSI device entries and test on them

The motivation for doing this was to reproduce the problem addressed in https://github.com/systemd/systemd/pull/8667 in the test suite, and to test for performance regressions (lock contention) caused by my proposed fix for the same issue.
I verified that all tests are still correct and pass. I actually made the test suite a bit harder to pass by adding additional consistence checks. The ability to test several symlinks in a single test allowed me to merge a few tests.
Two of the tests added in the "add multiple device test" commit do **FAIL** sporadically with the current code because of the mentioned race condition.

### CODE REVIEW

File: test/udev-test.pl [hunk 1]
```diff
@@ -43,52 +43,75 @@
...
+                devices => [
+                        {
+                                devpath         => "/devices/pci0000:00/0000:00:1f.2/host0/target0:0:0/0:0:0:0/block/sda",
+                                exp_name        => "sda" ,
+                                exp_rem_error   => "yes",
+                        },
+                        {
+                                devpath         => "/devices/pci0000:00/0000:00:1f.2/host0/target0:0:0/0:0:0:0/block/sda/sda1",
+                                exp_name        => "sda1" ,
+                                exp_rem_error   => "yes",
```
Comment (keszybz) [inline test/udev-test.pl]: So this one is actually the single (?) case where `devices` has two items. It'd be great to make this a separate commit, so it is not buried in all the other trivial cases where `devices` has one item.

File: test/udev-test.pl [hunk 2]
```diff
@@ -2353,7 +2366,7 @@ sub run_test {
 }
 
 $sema->remove;
-print "$error errors occurred\n\n";
+print "$error errors occurred. $good good results.\n\n";
```
Comment (keszybz) [inline test/udev-test.pl]: I think it would make sense to go one step further, and verify that `good` is above some threshold (and that threshold would have to be adjusted when tests are added or removed).

**Comment by mwilck** [comment]
> It seems we need to tweak our CI, otherwise we can't even check this properly.

Oops - IPC::SysV is a "core" Perl module, I didn't expect that to cause problems. That's not the only module I needed to add; there's also POSIX, IPC:Semaphore, Time:HiRes, and Cwd. They are all core modules.

**Comment by keszybz** [comment]
I added IPC::SysV to Fedora CI in https://src.fedoraproject.org/rpms/systemd/c/1eb8bc72df8f4dccff1e094b11eefce8adda896f?branch=rpm-master. The other ones might be pulled in by something else. Let's add them if it turns out that they are missing. If you push again, Fedora CI will run.

The failures on Ubuntu seem to be caused by something different.
```
autopkgtest [21:19:55]: test udev: [-----------------------
umount: test/tmpfs: no mount point specified.
Can't exec "/usr/bin/mknod": No such file or directory at ./udev-test.pl line 2254.
...
91 errors occurred. 6539 good results.
```

**Comment by mwilck** [comment]
> The tests fail for me after 2d27a67 'add multiple device test' (sometimes), and then the next commit is d299247 'add repeat count' after which the tests fail reliably.

Right. That's because these tests show a deficiency of the current udev code. The failure is sporadic, therefore adding the repeat count increases the failure probability. Merging https://github.com/systemd/systemd/pull/8667 should fix this. I kept the two pull requests separate on purpose for now.

> umount: test/tmpfs: no mount point specified.

That happens all the time. AFAIK it's not caused by my patches, but I'll double-check.

> Can't exec "/usr/bin/mknod": No such file or directory at ./udev-test.pl line 2254

OK, it's /bin/mknod on Ubuntu. I'll fix that.

**Comment by mwilck** [comment]
I just pushed fixes for the mknod and umount issues you mentioned.

Wrt the "multiple-devices" test, I need to understand the policy. If all tests are required to pass do to CI requirements, I can add "`exp_err_`" flags to these tests. We'd need to mark clearly that these errors are only "expected" because we know the current udev code is buggy.

File: test/udev-test.pl (hunk 2 above)
Comment (mwilck) [reply test/udev-test.pl]: I added an additional patch for the "expected good" count.

**Comment by mwilck** [comment]
@keszybz, anything else I can do to get this merged?

**Comment by evverx** [comment]
> Oops - IPC::SysV is a "core" Perl module, I didn't expect that to cause problems.

`Fedora`, `CentOS` and `RH` are unusual in the sense that an additional package has to be installed there to get the core Perl modules:
* https://fedoraproject.org/wiki/Changes/perl_Package_to_Install_Core_Modules
* https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/7/html/Developer_Guide/lib.details.html#libraries.perl

, so in practice, any use of any module might cause the test to fail immediately. I think it would make sense to check if all the modules used in `test-udev.pl` are installed and skip the test if they aren't.

> Two of the tests added in the "add multiple device test" commit do FAIL sporadically with the current code because of the mentioned race condition.

Would it be better to move the test cases related to https://github.com/systemd/systemd/pull/8667 there so that the addition of the expected errors could be avoided?

**Comment by mwilck** [comment]
>  I think it would make sense to check if all the modules used in test-udev.pl are installed and skip the test if they aren't.

That's dangerous for a regression test. The current behavior - fail the test, and require one-time developer action in the CI setup - seems to be safer to me. For the IPC::SysV module, that has happened already according to @keszybz' comment above, so AFAICS there's no further action needed, right?

> Would it be better to move the test cases related to #8667 there so that the addition of the expected errors could be avoided?

I could do that. I can also disable the tests, or modify the test to temporarily ignore certain failures and spit out a warning. I was wondering whether systemd might have an established policy for cases like this. 

What exactly is the recommendation?

**Comment by evverx** [comment]
> That's dangerous for a regression test.

I think it depends on the perception of the test. If it's expected that it has to fail no matter what, then checks like b1ffacb631d2209174 should be removed as well.

> For the IPC::SysV module, that has happened already according to @keszybz' comment above, so AFAICS there's no further action needed, right?

As far as I can see, `perl(IPC::SysV)` was added as a build dependency. It is a great shortcut, which made the test work on `Fedora Rawhide CI x86_64`. To make it more robust, though, `perl` should be installed as described in https://fedoraproject.org/wiki/Changes/perl_Package_to_Install_Core_Modules.

> What exactly is the recommendation?

In my opinion this PR is just a part of https://github.com/systemd/systemd/pull/8667, so I think it would make sense to move all the commits there and continue the discussion there too. It would automatically resolve all the issues related to whether the test should be disabled due to the expected failures or not.

**Comment by mwilck** [comment]
Superseded by https://github.com/systemd/systemd/pull/9551.
